lambda/matching.ml

```diff
@@ -690,6 +683,70 @@ end = struct
       | _ ->
           pretty_matrix Format.err_formatter pss;
           fatal_error "Matching.Default_environment.specialize_matrix"
+
+    (* Filter just one row, without a `rem` accumulator
+       of further rows to process.
+       The following equality holds:
+         filter_rec ((p :: ps) :: rem)
+         = filter_one p ps @ filter_rec rem
+    *)
+    and filter_one p ps =
+      filter_rec [ p :: ps ]
+
+    and filter_rec_or p1 p2 ps rem =
+      match arity with
+      | 0 -> (
+          (* if K has arity 0, specializing ((K|K)::rem) returns just (rem):
+             if either sides works (filters into a non-empty list),
+             no need to keep the other. *)
+          match filter_one p1 ps with
+          | [] -> filter_rec ((p2 :: ps) :: rem)
+          | matches -> matches @ filter_rec rem
+        )
```

Comment:
For ulterior reference, in the `matcher=pop_column` case, this is a deoptimization: we are splitting the or-pattern before dropping it once we call the matcher rather than dropping it from the outset.

Reply:
This is not how I read the code. For example if we consider an input pattern of the form `Tpat_or(p1, p2)` where `p1` is a simple pattern, then `filter_rec ((Tpat_or(p1, p2) :: ps) :: rem` is going to run `filter_one p1 ps`; because `p1` is simple this calls `matcher p1 ps` which  returns `[ps]` (with the `pop_column` matcher), and thus the code returns `ps :: filter_rec rem`. This is equivalent to what the previous code did, with the same number of calls to the matcher (on `p1` instead of `Tpat_or(p1,p2)` before) and same output. I wouldn't say we are splitting the row -- which in my mind implies a row duplication, in the output or at least in rows processed by intermediate computations.

Reply:
I think that we agree? Maybe exploring the `or-pattern` would be clearer. But there are more call to `filter_rec` (with deep or-patterns)  but still one call to the matcher, and the output will be same.
I doubt that the difference matters in practice.

Reply:
Ah, I thought you meant that the matcher would be called on both subpatterns, and/or that the row may end up twice in the result. You mean that it will traverse the pattern to the first simple subpattern at compilation time.

Reply:
Indeed.